qubes-trust-daemon.cpp

```diff
@@ -0,0 +1,355 @@
+    // Return if given empty set
+    if (file_paths.size() <= 0) {
+        return;
+    }
+
+    pid_t child_pid;
+    int exit_code;
+
+    // Convert set to array
+    const char* qvm_argv[MAX_ARG_LEN + 3]; // Account for extra args
+
+    // Add non-variable program arguments
+    qvm_argv[0] = "qvm-file-trust";
+    qvm_argv[1] = "--untrusted";
+
+    // Iterate through file path set and add to argv of qvm-file-trust
+    const char* file_path;
+    
+    // Loop through list of arguments, processing MAX_ARG_LEN args each time
+    int iterations = file_paths.size() / MAX_ARG_LEN;
+    std::set<std::string>::iterator it = file_paths.begin();
+    for (int i = 0; i <= iterations; i++) {
+        int arg_index = 2;
+
+        // Build an array of MAX_ARG_LEN elements
+        for (; arg_index - 2 < MAX_ARG_LEN; ++it) {
+            // Stop when we've hit the end
+            if (it == file_paths.end())
+                break;
+
+            file_path = (*it).c_str();
+
+            printf("Marking untrusted:: %s\n", file_path);
+
+            qvm_argv[arg_index] = file_path;
+            arg_index++;
+        }
+
+        // Add NULL terminator to signal end of argument list
+        qvm_argv[arg_index] = (char*) NULL;
+
+        // Once we have a list of MAX_ARG_LEN args,
+        // fork and attempt to call qvm-file-trust
+        switch (child_pid=fork()) {
+            case 0:
+                // We're the child, call qvm-file-trust
+                execv("/usr/bin/qvm-file-trust", (char**) qvm_argv);
+
+                // Unreachable if no error
+                perror("execl qvm-file-trust");
+                exit(1);
+            case -1:
+                // Fork failed
+                perror("fork failed");
+                exit(1);
+            default:
+                // Fork succeeded, and we got our pid, wait until child exits
+                if (waitpid(child_pid, &exit_code, 0) == -1) {
+                    perror("wait for qvm-file-trust failed");
+                    exit(1);
+                }
+        }
+    }
+
+    untrusted_buffer.clear();
+}
+
+/*
+ * Places an inotify_watch on the directory and all subdirectories
+ * Should call above method after compiling a list of all files to set as untrusted
+ */
+int watch_dir(const char *filepath, const struct stat *info,
+              const int typeflag, struct FTW *pathinfo) {
+    // Watch directories, set files as untrusted
+	struct stat s;
+	if(stat(filepath, &s) == 0) {
+		if(!(s.st_mode & S_IFDIR)) {
+            // File, set as untrusted
+            untrusted_buffer.insert(filepath);
+			return 0;
+		}
+	}
+	else {
+		// Error reading
+		return 1;
+	}
+
+    std::cout << "Placing watch on " << filepath
+        << " and subdirectories" << std::endl;
+
+    // Add watch to starting directory
+    int wd = inotify_add_watch(watch_fd, filepath, 
+		IN_CREATE | IN_MODIFY | IN_DELETE | IN_MOVED_TO | IN_MOVED_FROM);
```

Comment:
MOVED_FROM?

Reply:
This is to remove the inotify watch on an directory once it's moved from an untrusted one, since we don't consider it trusted anymore. Currently we don't remove the watch recursively though, leading to an error when a file in created in a sub-directory of a folder that was moved out of an untrusted folder.

Doesn't crash the daemon or anything, but definitely something I'll fix up.